### PR TITLE
Ajout des exercices pour la formation Machine Learning

### LINKED ISSUE #2: Création de la formation Machine Learning

### PR DESCRIPTION
**Pull request by fabricerenard12**
# Objectif
Resolves #2 
Cette PR introduit un notebook interactif destiné à accompagner la formation en Machine Learning. Il a pour but de consolider les notions fondamentales du ML à travers des exercices pratiques en Python, en s’appuyant sur des bibliothèques reconnues telles que scikit-learn, numpy et pytorch.

# Contenu ajouté
- Implémentations guidées de modèles classiques :
  - Régression linéaire & polynomiale
  - Régression logistique
  - Arbres de décision, forêts aléatoires, gradient boosting
  - Réseaux de neurones simples
- Séparation train/test, entraînement et évaluation
- Visualisations et interprétation des résultats

La PR inclut aussi une section Pour aller plus loin, regroupant des liens vers des cours, vidéos, notes et ouvrages de référence pour approfondir selon son niveau.

# À vérifier
- Exécuter tous les blocs du notebook (pas d’erreur)
- Vérifier que toutes les bibliothèques requises sont listées
- Orthographe et clarté des consignes

### CODE REVIEW

File: ML/Formation_ML.ipynb [hunk 1]
```diff
@@ -0,0 +1,1364 @@
...
+      "cell_type": "code",
+      "source": [
+        "def generate_polynomial_data(n_samples: int, noise: int, random_state: int = 0) -> Tuple[np.ndarray, np.ndarray]:\n",
+        "    \"\"\"\n",
+        "    Génère un dataset suivant une courbe polynomiale\n",
+        "    \"\"\"\n",
+        "    rng = np.random.default_rng(random_state)\n",
+        "    X = rng.uniform(-2, 2, size=(n_samples, 1))\n",
+        "    x = X[:, 0]\n",
+        "    y = 3 * x**3 - 2 * x**2 + x + noise * rng.normal(size=n_samples)\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:461]: Expliquer d'où vient la formule et quelques sont les paramètre aka x which is okay, mais noise et rng.normal, c'est serait bien d'avoir des précisions (biais)

File: ML/Formation_ML.ipynb [hunk 2]
```diff
@@ -0,0 +1,1364 @@
...
+        "    \"\"\"\n",
+        "    # TODO: Implémenter l'entraînement du modèle de régression polynomiale\n",
+        "\n",
+        "    raise NotImplementedError()\n",
+        "\n",
+        "def eval_poly(model: LinearRegression, poly: PolynomialFeatures, X_val: np.ndarray, y_val: np.ndarray) -> dict[str, float]:\n",
+        "    \"\"\"\n",
+        "    Évalue le modèle de régression polynomiale sur l'ensemble de validation transformé et renvoie les métriques.\n",
+        "    \"\"\"\n",
+        "    # TODO: Implémenter la validation du modèle de régression polynomiale\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:518]: précision des clés again

File: ML/Formation_ML.ipynb [hunk 3]
```diff
@@ -0,0 +1,1364 @@
...
+        "\n",
+        "def main(test_ratio: float = 0.2, random_state: int = 0, degree: int = 3):\n",
+        "    X, y = generate_polynomial_data(n_samples=500, noise=3)\n",
+        "\n",
+        "    # TODO: Séparer le jeu de données en ensembles d'entraînement et de validation\n",
+        "\n",
+        "    # TODO: Entraînement et évaluation du modèle de régression linéaire\n",
+        "\n",
+        "    # TODO: Entraînement et évaluation du modèle de régression polynomiale\n",
+        "\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:546]: Ça serait bien d'être uniforme dans les todos aka ce ce n'est pas écrit comme ça dans la partie régression linéaire

File: ML/Formation_ML.ipynb [hunk 4]
```diff
@@ -0,0 +1,1364 @@
...
+        "id": "Wz3S_HsRBLSr"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "class NeuralNet(nn.Module):\n",
+        "    # TODO: Implémenter l'architecture du réseau de neurones\n",
+        "    def __init__(self):\n",
+        "        raise NotImplementedError()\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:969]: explicité la séquence à faire (préférable)

File: ML/Formation_ML.ipynb [hunk 5]
```diff
@@ -0,0 +1,1364 @@
...
+        "\n",
+        "Apprendre les étapes clés d'un workflow de deep learning :\n",
+        "\n",
+        "- Charger et transformer les données MNIST (`transforms.ToTensor`)\n",
+        "- Construire un réseau de neurones avec `torch.nn.Module`\n",
+        "- Entraîner (`train`) et évaluer (`eval`) le modèle avec `CrossEntropyLoss` et un optimiseur (`Adam`)\n",
+        "- Mesurer les performances (Accuracy)\n",
+        "- Visualiser les prédictions sur des exemples d'images\n",
+        "\n",
+        "Vous implémenterez les fonctions `train`, `eval` et `main` pour réaliser ce processus de bout en bout.\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:907]: entrainement selon un nombre epoch?

File: ML/Formation_ML.ipynb [hunk 6]
```diff
@@ -0,0 +1,1364 @@
...
+        "\n",
+        "    transform = transforms.ToTensor()\n",
+        "\n",
+        "    train_ds = datasets.MNIST(root=\"data\", train=True, download=True, transform=transform)\n",
+        "    test_ds = datasets.MNIST(root=\"data\", train=False, download=True, transform=transform)\n",
+        "\n",
+        "    train_loader = DataLoader(train_ds, batch_size=batch_size, shuffle=True,  num_workers=2)\n",
+        "    test_loader = DataLoader(test_ds,  batch_size=batch_size, shuffle=False, num_workers=2)\n",
+        "\n",
+        "    # TODO: Instancier votre modèle, votre perte et votre optimiseur\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:999]: ça serait bien repriser la fonction de perte et l'optimiser à utiliser ici

**Review by nguyenm31 — COMMENTED** [review]
great job!

File: ML/Formation_ML.ipynb [hunk 7]
```diff
@@ -0,0 +1,1668 @@
...
+        "    degree : int, optional\n",
+        "        Degré du polynôme à utiliser pour la transformation des données\n",
+        "        (par défaut 2).\n",
+        "\n",
+        "    Returns\n",
+        "    -------\n",
+        "    model : LinearRegression\n",
+        "        Modèle de régression linéaire ajusté sur les caractéristiques polynomiales.\n",
+        "    poly : PolynomialFeatures\n",
+        "        Transformateur PolynomialFeatures utilisé pour générer les\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:521]: Transformateur utilisé pour générer...

File: ML/Formation_ML.ipynb [hunk 8]
```diff
@@ -0,0 +1,1668 @@
...
+        "    \"\"\"\n",
+        "    Évalue les performances du modèle sur un ensemble de validation.\n",
+        "\n",
+        "    Parameters\n",
+        "    ----------\n",
+        "    model : LinearRegression\n",
+        "        Modèle de régression linéaire préalablement entraîné.\n",
+        "    X_val : np.ndarray\n",
+        "        Matrice des variables explicatives de validation,\n",
+        "        de forme (n_samples, n_features).\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:223]: Matrice des variables observées de forme (n_samples, n_features) pour la validation

File: ML/Formation_ML.ipynb [hunk 9]
```diff
@@ -0,0 +1,1668 @@
...
+        "    y_val : np.ndarray\n",
+        "        Vecteur des valeurs cibles de validation,\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:225]: Vecteur des valeurs cibles pour la validation,\

File: ML/src/logistic_regression.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    Le modèle utilise le solver 'lbfgs' avec un nombre maximal d'itérations fixé à 1000
+    pour assurer la convergence.
+
+    Parameters
+    ----------
+    X_train : np.ndarray
+        Matrice des variables explicatives d'entraînement,
+        de forme (n_samples, n_features).
+    y_train : np.ndarray
+        Vecteur des labels cibles d'entraînement,
```
Comment (nguyenm31) [inline ML/src/logistic_regression.py:29]: Vecteur des classes cibles d'entraînement,

File: ML/Formation_ML.ipynb [hunk 11]
```diff
@@ -0,0 +1,1668 @@
...
+      "source": [
+        "DEVICE = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
+        "\n",
+        "class NeuralNet(nn.Module):\n",
+        "    \"\"\"\n",
+        "    Réseau de neurones entièrement connecté (MLP) pour la classification MNIST.\n",
+        "\n",
+        "    Architecture :\n",
+        "    - nn.Flatten() : aplatissement des images 28x28 en vecteurs 1D de taille 784.\n",
+        "    - Couche linéaire 1 : 784 → 128 neurones.\n",
```
Comment (nguyenm31) [inline ML/Formation_ML.ipynb:1120]: enlever lineaire
